src/index/update.rs

```diff
@@ -174,54 +184,64 @@ impl IndexUpdater {
 				continue;
 			}
 
-			if let Some(file_path_string) = file_path.to_str() {
-				if let Some(tags) = metadata::read(file_path.as_path()) {
-					if tags.year.is_some() {
-						inconsistent_directory_year |=
-							directory_year.is_some() && directory_year != tags.year;
-						directory_year = tags.year;
-					}
-
-					if tags.album.is_some() {
-						inconsistent_directory_album |=
-							directory_album.is_some() && directory_album != tags.album;
-						directory_album = tags.album.as_ref().cloned();
-					}
-
-					if tags.album_artist.is_some() {
-						inconsistent_directory_artist |=
-							directory_artist.is_some() && directory_artist != tags.album_artist;
-						directory_artist = tags.album_artist.as_ref().cloned();
-					} else if tags.artist.is_some() {
-						inconsistent_directory_artist |=
-							directory_artist.is_some() && directory_artist != tags.artist;
-						directory_artist = tags.artist.as_ref().cloned();
-					}
-
-					let song = NewSong {
-						path: file_path_string.to_owned(),
-						parent: path_string.to_owned(),
-						disc_number: tags.disc_number.map(|n| n as i32),
-						track_number: tags.track_number.map(|n| n as i32),
-						title: tags.title,
-						duration: tags.duration.map(|n| n as i32),
-						artist: tags.artist,
-						album_artist: tags.album_artist,
-						album: tags.album,
-						year: tags.year,
-						artwork: artwork.as_ref().cloned(),
-					};
-
-					self.push_song(song)?;
-				}
+			song_files.push(file_path);
+		}
+
+		let song_metadata = |path: PathBuf| -> Option<(String, SongTags)> {
+			#[cfg(feature = "profile-index")]
+			let _guard = flame::start_guard("song_metadata");
+
+			path.to_str().and_then(|file_path_string| {
+				metadata::read(&path).map(|m| (file_path_string.to_owned(), m))
+			})
+		};
+		let song_tags = song_files
+			.into_par_iter()
+			.filter_map(song_metadata)
+			.collect::<Vec<_>>();
```

Comment:
We could avoid the allocation here by doing a `fold`.

Reply:
Good enough for me and very readable in this form